streams/src/main/java/org/apache/kafka/streams/state/TimestampedWindowStore.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state;
+
+import org.apache.kafka.streams.processor.StateStore;
+
+/**
+ * A windowed store interface extending {@link StateStore}.
+ *
+ * In contrast to a {@link WindowStore} that stores plain key-value pairs,
+ * a {@code TimestampedWindowStore} stores key-(value/timestamp) pairs.
+ *
+ * @param <K> Type of keys
+ * @param <V> Type of values
+ */
+public interface TimestampedWindowStore<K, V> extends WindowStore<K, ValueAndTimestamp<V>> {
+
+    /**
+     * Put a key-value pair and its timestamp into the window with given window start timestamp
+     * @param key The key to associate the value to
+     * @param value The value, can be {@code null};
+     *              if the serialized bytes are also {@code null} it is interpreted as deletes
+     * @param timestamp the timestamp of the key-value pair; ignored if {@code value} is {@code null}
+     * @param windowStartTimestamp The timestamp of the beginning of the window to put the key/value/timestamp into
+     * @throws NullPointerException if the given key is {@code null}
+     */
+    void put(K key, V value, long timestamp, long windowStartTimestamp);
```

Comment:
This is a meta comment: I just realized that we are effectively having a duplicate API for each `put` operation in the extended class, like:

```
TimestampedWindowStore#put(K, ValueAndTimestamp<V>, long);
```

and

```
TimestampedWindowStore#put(K, V, long, long);
```

similarly for other two classes' put operations. I understand the first one is inherited and the second one is for more efficient store access, since there is often no reason to create a `ValueAndTimestamp` object if I already have a `V` value and a `long` timestamp at hand in my code. This duplicate API still degraded the API cleanness a bit: for example, for a customized TimestampedXXStore I need to implement two duplicated APIs still.

I'm wondering if there is a good way to go around it. If there's no better way at least we should consider just explicitly extend e.g. `put(K, ValueAndTimestamp<V>, long)` and default impl to the other API, and also add javadocs indicating that they have exactly the same semantics, and users should consider use the other one instead.

WDYT @bbejeck @vvcephei @ableegoldman ?

Reply:
As discussed in person, I think adding a default impl that "unwraps" the `ValueAndTimestamp` parameter into plain `V value` and `long timestamp` that calls the other/plain `put()` method seems to be a good idea.

Reply:
Huh, I previously overlooked this. Why do we need these new methods in addition to the equivalent ones we already have? IIRC, we agreed on `ValueAndTimestamp` to begin with to avoid having to create new store interfaces entirely.

Reply:
As discussed and announced in the KIP. Removed those methods.